File: coding/src/com/algos/li25_greedy/L179_LargestNumber.java
package edu.algos.li25_greedy;

import java.util.Arrays;

public class L179_LargestNumber {

	public static void main(String[] args) {
		L179_LargestNumber obj = new L179_LargestNumber();
		
		int[] nums = new int[] {3,30,34,5,9};
		System.out.println(obj.largestNumber(nums));
	}

	/*
	 * Problem: https://leetcode.com/problems/largest-number/
	 * Companies: Amazon
	 * Idea: Convert the nums to str and use str sorting
	 * */
	public String largestNumber(int[] nums) {
        String[] numsStr = new String[nums.length];
		for(int i=0; i<nums.length; i++) {
			numsStr[i] = String.valueOf(nums[i]);
		}
		Arrays.sort(numsStr, (n1, n2) -> (n2+n1).compareTo((n1+n2)));
        
        if(numsStr[0].equals("0")) return "0";
        return String.join("", numsStr);
    }
}

File: dsa/topics/21_greedy/python/l179_largest_number.py
from functools import cmp_to_key

# Problem: https://leetcode.com/problems/largest-number/
# Companies: Amazon
# Idea: Convert the nums to str and use str sorting

def largestNumber(nums):
    nums_str = list(map(str, nums))

    # Custom comparator: decide order by comparing concatenations
    def compare(a, b):
        if a + b > b + a:
            return -1
        if a + b < b + a:
            return 1
        return 0

    nums_str.sort(key=cmp_to_key(compare))

    # Edge case: when all numbers are zero
    if nums_str[0] == "0":
        return "0"

    return "".join(nums_str)


# -------- Test --------
print(largestNumber([3, 30, 34, 5, 9]))  # Expected: "9534330"
